src/stan/model/indexing/lvalue_varmat.hpp

```diff
@@ -0,0 +1,396 @@
+ *       over the column subset.
+ * Std vector:
+ *  - single element and elementwise overloads
+ *  - General overload for nested std vectors.
+ */
+
+/**
+ * Assign to a single element of an Eigen Vector.
+ *
+ * Types: vector[uni] <- scalar
+ *
+ * @tparam Vec `var_value` with inner Eigen type with either dynamic rows or
+ * columns, but not both.
+ * @tparam U Type of value (must be assignable to T).
+ * @param[in] x Vector variable to be assigned.
+ * @param[in] idxs index holding which cell to assign to.
+ * @param[in] y Value to assign.
+ * @param[in] name Name of variable (default "ANON").
+ * @param[in] depth Indexing depth (default 0).
+ * @throw std::out_of_range If the index is out of bounds.
+ */
+template <typename VarVec, typename U, require_var_vector_t<VarVec>* = nullptr,
+          require_var_t<U>* = nullptr,
+          require_floating_point_t<value_type_t<U>>* = nullptr>
+inline void assign(VarVec&& x,
+                   const cons_index_list<index_uni, nil_index_list>& idxs,
+                   const U& y, const char* name = "ANON", int depth = 0) {
+  stan::math::check_range("var_vector[uni] assign", name, x.size(),
+                          idxs.head_.n_);
+  const auto coeff_idx = idxs.head_.n_ - 1;
+  double prev_val = x.val().coeffRef(coeff_idx);
+  x.vi_->val_.coeffRef(coeff_idx) = y.val();
+  stan::math::reverse_pass_callback([x, y, coeff_idx, prev_val]() mutable {
+    x.vi_->val_.coeffRef(coeff_idx) = prev_val;
+    prev_val = x.adj().coeffRef(coeff_idx);
+    x.adj().coeffRef(coeff_idx) = 0.0;
+    y.adj() += prev_val;
+  });
+}
+
+/**
+ * Assign to a non-contiguous subset of elements in a vector.
+ *
+ * Types:  vector[multi] <- vector
+ *
+ * @tparam Vec1 `var_value` with inner Eigen type with either dynamic rows or
+ * columns, but not both.
+ * @tparam Vec2 `var_value` with inner Eigen type with either dynamic rows or
+ * columns, but not both.
+ * @param[in] x Vector to be assigned.
+ * @param[in] idxs Index holding an `std::vector` of cells to assign to.
+ * @param[in] y Value vector.
+ * @param[in] name Name of variable (default "ANON").
+ * @param[in] depth Indexing depth (default 0).
+ * @throw std::out_of_range If any of the indices are out of bounds.
+ * @throw std::invalid_argument If the value size isn't the same as
+ * the indexed size.
+ */
+template <typename Vec1, typename Vec2,
+          require_all_var_vector_t<Vec1, Vec2>* = nullptr>
+inline void assign(Vec1&& x,
+                   const cons_index_list<index_multi, nil_index_list>& idxs,
+                   const Vec2& y, const char* name = "ANON", int depth = 0) {
+  stan::math::check_size_match("vector[multi] assign", "left hand side",
+                               idxs.head_.ns_.size(), name, y.size());
+  const auto x_size = x.size();
+  const auto assign_size = idxs.head_.ns_.size();
+  arena_t<std::vector<int>> x_idx;
+  arena_t<std::vector<int>> y_idx;
+  // Need to remove duplicates for cases like {2, 3, 2, 2}
```

Comment:
When I was working through this, if we have something like 

```stan
vector[3] x;
vector[6] y;
x[{1, 2, 3, 1, 2, 3}] = y;
```

Since the `Matrix<var>` effectively just assigns a pointer this is effectivly doing a de-dup and taking the last value assigned like

```
x = y[{4, 5, 6}];
```

If we don't de-dup in `var<mat>` then the y indices `{1, 2, 3}` would differ between the `mat<var>` and `var<mat>` version, where in the `mat<var>` version it's as if they were never touched but in the `var<mat>` version we would end up propagate up those adjoints. Does that make sense? I'd love to be full of baloney on this one but I think we need to do this de-duping scheme here

I've been trying to think of smarter ways to do this but I keep getting stuck at how to get out the proper `y` values

Reply:
I see the issue. But that quadratic deduplication will hurt too much on large matrices. Deduplication can be done in linear time if you do duplicate checks on `std::unordered_set` instead of `std::vector`. To improve memory access pattern you should build that set in the same loop you are storing indices and do checks on it in the reverse pass loop.

Reply:
Do you mean something like this?

```
  std::unordered_set<int, std::hash<int>, std::equal_to<int>, stan::math::arena_allocator<int>> x_set;
  arena_t<std::vector<int>> x_idx;
  arena_t<std::vector<int>> y_idx;
  x_set.reserve(assign_size);
  x_idx.reserve(assign_size);
  y_idx.reserve(assign_size);
  // Need to remove duplicates for cases like {2, 3, 2, 2}
  for (int i = assign_size - 1; i >= 0; --i) {
    if (x_set.insert(idxs.head_.ns_[i]).second) {
      stan::math::check_range("vector[multi] assign", name, x_size,
                              idxs.head_.ns_[i]);
      y_idx.push_back(i);
      x_idx.push_back(idxs.head_.ns_[i] - 1);
    }
  }
```

else we could store them in a regular set and not have to have the extra mem (I think?)

Reply:
Yeah I meant like this. Or maybe we should instead just document it that duplicating indices on left hand side of assignment is not supported.

Reply:
I think we should do the unordered_set thing for now. I'll make an issue in stanc3 a `unique_index` type that would just be an `std::vector` but checks at compilation that each element is unique. Then we can have a specialization to handle unique index that doesn't need to do the de-dup